functests/gatekeeper/gatekeeper.go

```diff
@@ -0,0 +1,637 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	dynamicInterface "github.com/openshift-kni/cnf-features-deploy/functests/utils/dynamicinterface"
+)
+
+type metadataInfo struct {
+	kind  string
+	key   string
+	value string
+}
+
+type assignMetadataInfo struct {
+	name     string
+	kinds    []string
+	scope    string
+	metadata metadataInfo
+}
+
+var _ = Describe("gatekeeper", func() {
+	gvrList := []schema.GroupVersionResource{
+		{
+			Group:    "mutations.gatekeeper.sh",
+			Version:  "v1alpha1",
+			Resource: "assignmetadata",
+		},
+	}
+
+	objects := GetAllTestObjects(TestingNamespace)
+	for _, object := range objects {
+		gvrList = append(gvrList, object.GroupVersionResource())
+	}
+
+	// create the dynamic client
+	client, err := dynamicInterface.New("", gvrList...)
+	if err != nil {
+		log.Fatalf("Unable to create the dynamic client: %v", err)
+	}
+
+	DescribeTable("should be able to add metadata info(labels/annotations)",
+		func(amInfo assignMetadataInfo, object TestObject) {
+			amObject := assignMetadataDefinition(&amInfo)
+
+			amObject, err := client.Resource("assignmetadata").Create(context.Background(), amObject, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			defer func() {
+				err := client.Resource("assignmetadata").Delete(context.Background(), amObject.GetName(), metav1.DeleteOptions{})
+				Expect(err).ToNot(HaveOccurred())
+			}()
+
+			object.Unstructured, err = client.Resource(object.resourceName).Namespace(object.GetNamespace()).Create(context.Background(), object.Unstructured, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			value, ok, err := unstructured.NestedString(object.Object, "metadata", amInfo.metadata.kind, "mutated")
+			Expect(err).ToNot(HaveOccurred())
+			Expect(ok).To(BeTrue())
+			Expect(value).To(Equal("true"))
+
+			err = client.Resource(object.resourceName).Namespace(object.GetNamespace()).Delete(context.Background(), object.GetName(), metav1.DeleteOptions{})
+			Expect(err).ToNot(HaveOccurred())
+		},
+		createTableEntries("added", []metadataInfo{
+			{
+				kind:  "labels",
+				key:   "mutated",
+				value: "true",
+			},
+			{
+				kind:  "annotations",
+				key:   "mutated",
+				value: "true",
+			},
+		}...)...,
+	)
+
+	DescribeTable("should not mutate existing metadata info(labels/annotations)",
+		func(amInfo assignMetadataInfo, object TestObject) {
+			amDefinition := assignMetadataDefinition(&amInfo)
+
+			amDefinition, err := client.Resource("assignmetadata").Create(context.Background(), amDefinition, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+			defer func() {
+				err := client.Resource("assignmetadata").Delete(context.Background(), amDefinition.GetName(), metav1.DeleteOptions{})
+				Expect(err).ToNot(HaveOccurred())
+			}()
```

Comment:
We have quite a lot of these defers, so maybe we can do:

```
func deleteResource(resource string, name string) func() {
   return func() {
	err := client.Resource(resource).Delete(context.Background(), name, metav1.DeleteOptions{})
	Expect(err).ToNot(HaveOccurred())
  }
}

```
and then use it like:
`
defer deleteResource("assignmetadata", amDefinition.GetName())()()
`

With 19 defer statements we should save more then 50 lines of code.